modules/api/php/endpoints/candidate/visit/dicoms.class.inc

```diff
@@ -140,6 +158,141 @@ class Dicoms extends Endpoint implements \LORIS\Middleware\ETagCalculator
+     * A successful upload should return a 303 response toward this DICOM
+     * processes.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    private function _handlePOST(ServerRequestInterface $request): ResponseInterface
+    {
+        // *************************
+        // *** Permission checks ***
+        // *************************
+        $user = $request->getAttribute('user');
+        if (!$user->hasPermission('imaging_uploader')) {
+            return new \LORIS\Http\Response\JSON\Forbidden();
+        }
+
+        // *************************
+        // *** Input Validation ****
+        // *************************
+        $files = $request->getUploadedFiles();
+        if (count($files) !== 1) {
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                'There should be 1 and only 1 file uploaded per POST request.'
+            );
+        }
+        if (!isset($files['mriFile'])){
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                'The input key for the uploaded file should be mriFile.'
+            );
+        }
+        $uploadedfile = $files['mriFile'];
+
+        $data = $request->getParsedBody();
+
+        $inputcandid = $data['CandID'] ?? '';
+        $visitcandid = $this->_visit->getCandID();
+        if ($inputcandid != (string) $visitcandid) {
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                'CandID in POST data do not match URL.'
+            );
+        }
+
+        $pscid = $data['PSCID'] ?? '';
+        if (!\Candidate::candidateExists($visitcandid, $pscid)) {
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                'PSCID and CandID in POST data do not refer to the same candidate.'
+            );
+        }
+
+        $visitlabel = $data['Visit'] ?? null;
+        if ($visitlabel !== $this->_visit->getVisitLabel()) {
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                'Visit label in URL and POST data do not match.'
+            );
+        }
+
+        $isphantom = $data['IsPhantom'] ?? null;
+        if (!in_array($isphantom, ['true', 'false'])) {
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                'IsPhantom must be a the string "true" or "false".'
+            );
+        }
+
+        $filename    = $uploadedfile->getClientFilename();
+        $patientname = $pscid. "_" . $inputcandid . "_" . $visitlabel;
+        if (!preg_match("/^${patientname}/", $filename) && $isphantom !== 'true') {
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                "Filename must start with $patientname."
+            );
+        }
+
+        // *************************
+        // *** Upload file *********
+        // *************************
+        $incomingpath = \NDB_factory::singleton()
```

Comment:
Might be a good idea to use trim() here and so You don't end up with `\n` and white space if the user has the tags with the path not on the same line.